File: src/adapters/probe_lit.py
# probe_lit.py —— v0.11.9m 乾坤镜 LIT Python SDK 适配层
# 职责：在 LIT Client/Server 事件流中注入乾坤镜探针调用
# 依赖：_python_base.py, _payload_builders.py（零第三方依赖）

from adapters._python_base import make_adapter_init, instrumented
from adapters._payload_builders import (
    build_llm_event,
    build_tool_event,
    build_memory_event,
    _extract_tokens,
    _extract_finish_reason,
)


def _monkey_patch(probe):
    try:
        from mcp.client.session import ClientSession

        ClientSession.call_tool = instrumented(
            probe,
            "tool_call",
            lambda s, r, lat, a, k: build_tool_event(
                tool_name=a[0] if a else None,
                execution_ms=lat,
                tool_args=k.get("arguments") or {},
                tool_result=r,
                target_host="mcp_server",
                success=r is not None and not isinstance(r, Exception),
            ),
            {"layer_agent": {"agent_name": "mcp_client"}},
        )(ClientSession.call_tool)

        ClientSession.read_resource = instrumented(
            probe,
            "memory_retrieve",
            lambda s, r, lat, a, k: build_memory_event(
                query=a[0] if a else None,
                results=r if isinstance(r, (list, tuple)) else ([r] if r else None),
                latency_ms=lat,
                memory_type="mcp_resource",
                memory_store=k.get("uri") or (str(a[0]) if a else None),
            ),
            {"layer_agent": {"agent_name": "mcp_client"}},
        )(ClientSession.read_resource)

        from mcp.server import Server

        _orig_handle = getattr(Server, "handle_request", None)
        if _orig_handle:
            Server.handle_request = instrumented(
                probe,
                "llm_invoke",
                lambda s, r, lat, a, k: build_llm_event(
                    agent_name="mcp_server",
                    latency_ms=lat,
                    response=r,
                    kwargs={
                        "input_tokens": _extract_tokens(r, k)[0],
                        "output_tokens": _extract_tokens(r, k)[1],
                        "finish_reason": _extract_finish_reason(r, k),
                    },
                    target_host="mcp_client",
                ),
                {"layer_agent": {"agent_name": "mcp_server"}},
            )(_orig_handle)

    except ImportError:
        pass


init_lit_probe = make_adapter_init(_monkey_patch, "lit")

File: src/adapters/probe_lit.js
// probe_lit.js —— 0.11.9m 乾坤镜 LIT TypeScript/Node.js 适配层
// 职责：在 LIT Client/Server 事件流中注入乾坤镜探针调用
// 依赖：_js_base.js, probe_node.js（零 npm 依赖）

const crypto = require('crypto');
const { makeAdapterInit, intercept } = require('./_js_base');
const probe = require('../probe_node');

function _computeHash(text) {
  if (!text) return null;
  return crypto.createHash('sha256').update(String(text)).digest('hex').substring(0, 16);
}

function _getContentMaxLen() {
  const val = process.env.MING_CONTENT_MAX_LEN;
  if (val === undefined || val === '') return 2000;
  const v = parseInt(val, 10);
  if (isNaN(v)) return 2000;
  return v === 0 ? 0 : v;
}

function truncate(s, maxLen) {
  if (maxLen === undefined || maxLen === null) maxLen = _getContentMaxLen();
  if (maxLen === 0) return s;
  const str = String(s);
  if (str.length > maxLen) return str.substring(0, maxLen) + '...';
  return str;
}

function _patch(probe) {
  if (typeof globalThis.McpClient !== 'undefined') {
    const C = globalThis.McpClient;
    intercept(C.prototype, 'callTool', 'tool_call',
      (r, lat, args) => ({
        layer_agent: { step_id: 'unknown', session_id: 'unknown', agent_name: 'mcp_client' },
        layer_tool: { tool_name: args[0], tool_args: args[1] || {}, tool_result: JSON.stringify(r).slice(0, 500), execution_ms: lat },
        layer_network: { target_host: 'mcp_server', status_code: 200 }
      }),
      { layer_agent: { step_id: 'unknown', session_id: 'unknown', agent_name: 'mcp_client' } }
    );
    intercept(C.prototype, 'readResource', 'memory_retrieve',
      (r, lat, args) => ({
        layer_agent: { step_id: 'unknown', session_id: 'unknown', agent_name: 'mcp_client' },
        layer_memory: { memory_type: 'mcp_resource', query: String(args[0]), results_count: 1, latency_ms: lat }
      }),
      { layer_agent: { step_id: 'unknown', session_id: 'unknown', agent_name: 'mcp_client' } }
    );
  }

  if (typeof globalThis.McpServer !== 'undefined') {
    const S = globalThis.McpServer;
    const origHandle = S.prototype.handleRequest;
    S.prototype.handleRequest = async function(...args) {
      const start = Date.now();
      const result = await origHandle.apply(this, args);
      const latencyMs = Date.now() - start;

      // 发射 llm_invoke 事件
      probe.emit('llm_invoke', {
        layer_agent: { step_id: 'unknown', session_id: 'unknown', agent_name: 'mcp_server' },
        layer_llm: { model: 'unknown', input_tokens: 0, output_tokens: 0, latency_ms: latencyMs, cache_hit: false },
        layer_network: { target_host: 'mcp_client', status_code: 200 }
      });

      // 发射 llm_output 事件（如果有输出）
      const outputText = result && (result.content || result.text || (typeof result === 'string' ? result : null));
      if (outputText) {
        probe.emit('llm_output', {
          layer_agent: { step_id: 'unknown', session_id: 'unknown', agent_name: 'mcp_server' },
          layer_llm: {
            output_text: truncate(outputText),
            output_text_hash: _computeHash(outputText)
          },
          layer_network: { target_host: 'mcp_client' }
        });
      }

      return result;
    };
  }
}

module.exports = { initLitProbe: makeAdapterInit('lit', _patch) };
